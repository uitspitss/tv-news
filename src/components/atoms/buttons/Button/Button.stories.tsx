import React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';
import base from 'paths.macro';

import { Button, ButtonProps } from './Button';

const path = base.replace(
  /^.+\/(atoms|molecules|organisms|templates|pages)\/(.+)\//,
  '$1/$2',
);

export default {
  title: `${path}`,
  component: Button,
} as Meta;

const Template: Story<ButtonProps> = (args) => <Button {...args} />;

export const Default = Template.bind({});
Default.args = {
  children: 'Button',
  primary: false,
  secondary: false,
};

export const Primary = Template.bind({});
Primary.args = {
  children: 'Button',
  primary: true,
  secondary: false,
};

export const Secondary = Template.bind({});
Secondary.args = {
  ...Default.args,
  secondary: true,
};

export const Large = Template.bind({});
Large.args = {
  ...Default.args,
  primary: false,
  size: 'large',
};

export const Small = Template.bind({});
Small.args = {
  ...Default.args,
  primary: false,
  size: 'small',
};
