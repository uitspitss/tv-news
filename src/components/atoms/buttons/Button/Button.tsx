import React, {
  memo,
  ComponentPropsWithRef,
  forwardRef,
  ForwardRefExoticComponent,
  ReactNode,
} from 'react';
import tw, { styled } from 'twin.macro';

export type ButtonProps = {
  /**
   * class for twin.macro
   */
  className?: string;
  /**
   * children
   */
  children: ReactNode;
  /**
   * primary color
   */
  primary?: boolean;
  /**
   * secondary color
   */
  secondary?: boolean;
  /**
   * button size
   */
  size?: 'large' | 'small';
} & ComponentPropsWithRef<'button'>;

type Props = ButtonProps;

const Component: ForwardRefExoticComponent<Props> = forwardRef((props, ref) => {
  const {
    className,
    type = 'button',
    primary: _primary,
    secondary: _secondary,
    children,
    ...rest
  } = props;

  return (
    <button
      className={className}
      // eslint-disable-next-line react/button-has-type
      type={type}
      ref={ref}
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...rest}
    >
      {children}
    </button>
  );
});

const StyledComponent = styled(Component)`
  ${tw`text-base rounded border-none px-2 py-1`}

  /* color */
  ${({ primary }) => primary && tw`bg-primary-500 text-gray-200`}
  ${({ secondary }) => secondary && tw`bg-secondary-500 text-gray-200`}

  /* size */
  ${({ size }) => size === 'large' && tw`text-lg`}
  ${({ size }) => size === 'small' && tw`text-sm`}
`;

/**
 * Button
 */
export const Button: ForwardRefExoticComponent<ButtonProps> = memo(
  forwardRef((props, ref) => {
    return <StyledComponent {...props} ref={ref} />;
  }),
);
