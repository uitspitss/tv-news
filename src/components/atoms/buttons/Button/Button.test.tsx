import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';

import { Button } from './Button';

describe('Button', () => {
  test('snapshot test', () => {
    const { container } = render(<Button>button</Button>);
    expect(container).toMatchSnapshot();
  });

  describe('Behavior', () => {
    const clickFn = jest.fn();

    beforeEach(() => {
      jest.clearAllMocks();
      render(<Button onClick={clickFn}>button</Button>);
    });

    test('initial view', () => {
      expect(
        screen.getByRole('button', { name: 'button' }),
      ).toBeInTheDocument();
      expect(screen.getByText('button')).toBeInTheDocument();
    });

    test('click, then clickFn is called', () => {
      fireEvent.click(screen.getByRole('button', { name: 'button' }));
      expect(clickFn).toBeCalledTimes(1);
    });
  });
});
