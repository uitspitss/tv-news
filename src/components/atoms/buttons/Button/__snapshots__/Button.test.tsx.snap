// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button snapshot test 1`] = `
.emotion-0 {
  font-size: 1rem;
  line-height: 1.5rem;
  border-radius: 0.25rem;
  border-style: none;
  padding-left: 0.5rem;
  padding-right: 0.5rem;
  padding-top: 0.25rem;
  padding-bottom: 0.25rem;
}

<div>
  <button
    class="emotion-0 emotion-1"
    type="button"
  >
    button
  </button>
</div>
`;
